check that first arg to `panic!()` in const is `&str`

abonander opened the pull request:
closes #66693

~~TODO: regression test~~

cc @RalfJung for error message wording

rust-highfive commented:
r? @oli-obk

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1652883134) [(plain)](https://github.com/rust-lang/rust/commit/6412cbad193036602c9e77a0b368e64ee7834638/checks/1652883134/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_ty_utils v0.0.0 (/checkout/compiler/rustc_ty_utils)
    Checking rustc_mir v0.0.0 (/checkout/compiler/rustc_mir)
    Checking rustc_traits v0.0.0 (/checkout/compiler/rustc_traits)
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
error[E0277]: can't compare `&rustc_middle::ty::TyKind<'_>` with `rustc_middle::ty::TyKind<'_>`
   --> compiler/rustc_mir/src/transform/check_consts/validation.rs:865:75
    |
865 |                     if args[0].ty(&self.ccx.body.local_decls, tcx).kind() != TyKind::Str {
    |                                                                           ^^ no implementation for `&rustc_middle::ty::TyKind<'_> == rustc_middle::ty::TyKind<'_>`
    |
    = help: the trait `std::cmp::PartialEq<rustc_middle::ty::TyKind<'_>>` is not implemented for `&rustc_middle::ty::TyKind<'_>`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `rustc_mir`
error: could not compile `rustc_mir`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_codegen_ssa" "-p" "rustc_apfloat" "-p" "rustc_serialize" "-p" "rustc_fs_util" "-p" "rustc_middle" "-p" "rustc_feature" "-p" "rustc_query_system" "-p" "rustc_type_ir" "-p" "rustc_arena" "-p" "rustc_hir" "-p" "rustc_ast" "-p" "rustc_lexer" "-p" "rustc_symbol_mangling" "-p" "rustc_span" "-p" "rustc_incremental" "-p" "rustc_graphviz" "-p" "rustc_attr" "-p" "rustc_ast_pretty" "-p" "rustc_macros" "-p" "rustc_data_structures" "-p" "rustc_target" "-p" "rustc_session" "-p" "rustc_lint_defs" "-p" "rustc_index" "-p" "rustc_errors" "-p" "rustc_driver" "-p" "rustc_parse" "-p" "rustc_hir_pretty" "-p" "rustc_save_analysis" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_ast_passes" "-p" "rustc_plugin_impl" "-p" "rustc_lint" "-p" "rustc_parse_format" "-p" "rustc_trait_selection" "-p" "rustc_infer" "-p" "rustc_interface" "-p" "rustc_ast_lowering" "-p" "rustc_builtin_macros" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_privacy" "-p" "rustc_passes" "-p" "rustc_typeck" "-p" "rustc_mir_build" "-p" "rustc_ty_utils" "-p" "rustc_traits" "-p" "rustc_resolve" "-p" "rustc_error_codes" "-p" "rustc_mir" "-p" "coverage_test_macros" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:56
```

</details>

oli-obk commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -860,6 +860,16 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                     self.check_op(ops::Panic);
+
+                    // an internal `panic()` call must have at least one argument
+                    if args[0].ty(&self.ccx.body.local_decls, tcx).kind() != TyKind::Str {
```
I think this is not going to work, because the type is `&str`, and this check is for `str`. I think what you can do is to first match with a `ty::Ref(_, ty, _)` pattern to get the thing behind the reference, and then there's actually a convenience function `is_str()` that will do the `.kind() == ty::Str` for you.

abonander replied:
@oli-obk I believe this recurses through the projection list to find the root type; it's what @RalfJung recommended. I did see `.is_str()` but I think it may be in the wrong module because my IDE didn't suggest it; should I just elevate its visibility?

abonander replied:
This call goes through the `Place::projection_items()` to get the final type, I believe.

oli-obk replied:
yes, but the "final type" is a `&str`. The reason it goes through `Place::Projection_items()` is so that `panic!(*x.foo)` will give us whatever value is at the destination of `foo`, so e.g. if `x: SomeStruct` where `struct SomeStruct { foo: &&str }`.

oli-obk replied:
You can insert `trace!` statements to check the type and/or make your new `ops` struct have a `Ty<'tcx>` field allowing the error message to emit the type that was encountered. This may make debugging easier in case a test isn't behaving as you expect it to.

RalfJung replied:
> I think this is not going to work, because the type is &str, and this check is for str. I think what you can do is to first match with a ty::Ref(_, ty, _) pattern to get the thing behind the reference, and then there's actually a convenience function is_str() that will do the .kind() == ty::Str for you.

Oh, good catch. Currently the code requires the argument type to be `str` but we want it to be `&str`.

oli-obk commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -860,6 +860,16 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
+
+                    // an internal `panic()` call must have at least one argument
+                    if args[0].ty(&self.ccx.body.local_decls, tcx).kind() != TyKind::Str {
+                        self.tcx.sess.struct_span_err(
```
Please add a new struct in the `ops` module (same directory). You can copy paste `struct FnCallIndirect`, which has similar behaviour to what we want here (always hard error, no exceptions, no feature gates). Then you can `self.check_op(ops::YourStruct)` to handle it like we handle all other errors in this file.

abonander replied:
@oli-obk `ops::Panic` already exists, should I just add this check there?

oli-obk replied:
no, `ops::Panic` is about allowing `panic!()` in general, and is a feature gated error. This here is an error about the wrong argument type to `panic!()`, and should for now stay a hard error. Some day we may be more permissive, but it's an entirely independent error (as you can see by the fact that the message you are emitting here (which should stay as the message for the new `ops` struct), is different from the one in `ops::Panic`)

oli-obk reviewed (commented):
does this affect any tests? If not, please add a test in a test that has the `const_panic` feature activated.

abonander commented:
> does this affect any tests? If not, please add a test in a test that has the `const_panic` feature activated.

Yep, I was getting to that. I just wanted to post as draft for early feedback on the error message.

oli-obk commented:
> Yep, I was getting to that. I just wanted to post as draft for early feedback on the error message.

oh duh, I should have read your main PR message properly..

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1653021888) [(plain)](https://github.com/rust-lang/rust/commit/c6c1997687e82fd38c6068b0eee8058c5b6a5f21/checks/1653021888/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir v0.0.0 (/checkout/compiler/rustc_mir)
error: unused import: `TyKind`
  --> compiler/rustc_mir/src/transform/check_consts/validation.rs:14:71
   |
14 |     self, adjustment::PointerCast, Instance, InstanceDef, Ty, TyCtxt, TyKind, TypeAndMut,
   |
   |
   = note: `-D unused-imports` implied by `-D warnings`
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
error: aborting due to previous error

error: could not compile `rustc_mir`
error: could not compile `rustc_mir`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_target" "-p" "rustc_index" "-p" "rustc_macros" "-p" "rustc_interface" "-p" "rustc_traits" "-p" "rustc_infer" "-p" "rustc_graphviz" "-p" "rustc_trait_selection" "-p" "rustc_parse_format" "-p" "rustc_lexer" "-p" "rustc_codegen_llvm" "-p" "rustc_fs_util" "-p" "rustc_llvm" "-p" "rustc_ty_utils" "-p" "rustc_incremental" "-p" "rustc_attr" "-p" "rustc_passes" "-p" "rustc_mir_build" "-p" "rustc_apfloat" "-p" "rustc_arena" "-p" "rustc_ast_passes" "-p" "rustc_expand" "-p" "rustc_symbol_mangling" "-p" "rustc_privacy" "-p" "rustc_builtin_macros" "-p" "rustc_typeck" "-p" "rustc_ast_lowering" "-p" "rustc_resolve" "-p" "rustc_feature" "-p" "rustc_serialize" "-p" "rustc_hir_pretty" "-p" "rustc_save_analysis" "-p" "rustc_parse" "-p" "rustc_error_codes" "-p" "rustc_ast" "-p" "rustc_metadata" "-p" "rustc_middle" "-p" "rustc_type_ir" "-p" "rustc_query_system" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_hir" "-p" "rustc_data_structures" "-p" "rustc_errors" "-p" "rustc_lint_defs" "-p" "rustc_lint" "-p" "rustc_session" "-p" "rustc_span" "-p" "rustc_plugin_impl" "-p" "rustc_ast_pretty" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:56
```

</details>

RalfJung commented on compiler/rustc_mir/src/transform/check_consts/ops.rs:
```diff
@@ -360,6 +360,17 @@ impl NonConstOp for Panic {
+    fn build_error(&self, ccx: &ConstCx<'_, 'tcx>, span: Span) -> DiagnosticBuilder<'tcx> {
+        ccx.tcx
+            .sess
+            .struct_span_err(span, "only argument to `panic!()` in a const context must be `&str`")
```
```suggestion
            .struct_span_err(span, "argument to `panic!()` in a const context must have type `&str`")
```

abonander commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -857,9 +857,17 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
+
+                    // an internal `panic()` call must have exactly one argument
+                    // and evaluation of the `panic` op assumes the argument is `&str`
+                    match args[0].ty(&self.ccx.body.local_decls, tcx).kind() {
```
I'm assuming this indexing is okay since the call would otherwise have been rejected by typechecking long before we got here.

RalfJung replied:
`is_lang_panic_fn` returns true for three different functions... does it really make sense to treat them all the same here? `panic_fn` and `panic_str` can never have any other argument I think, by mere type-checking. So the test probably should only fire for `begin_panic_fn`.

abonander commented:
@RalfJung @oli-obk I adjusted the wording, fixed the check and added UI tests. I noticed a discrepancy between panics in array length expressions and normal consts so I put them in separate files. Should the tests cover `const fn` as well?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1665962394) [(plain)](https://github.com/rust-lang/rust/commit/a5e15c72a9b382de54f4e7d688c81268fec10115/checks/1665962394/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9000/11247
.................................................................................................... 9100/11247
.................................................................................................... 9200/11247
...........................................i......i................................................. 9300/11247
..................................................................................iiiiii..iiiiii.i.. 9400/11247
.................................................................................................... 9600/11247
.................................................................................................... 9700/11247
.................................................................................................... 9800/11247
.................................................................................................... 9900/11247
---
8    = help: add `#![feature(const_panic)]` to the crate attributes to enable
9    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
10 
- error: aborting due to previous error
+ error: argument to `panic!()` in a const context must have type `&str`
+    |
+    |
+ LL | struct Bug([u8; panic!(1)]);
+    |
+    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
+ 
+ error: aborting due to 2 previous errors
+ error: aborting due to 2 previous errors
12 
13 For more information about this error, try `rustc --explain E0658`.
14 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/issue-76064/issue-76064.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args consts/issue-76064.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/consts/issue-76064.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/issue-76064" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/issue-76064/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0658]: panicking in constants is unstable
  --> /checkout/src/test/ui/consts/issue-76064.rs:1:17
   |
LL | struct Bug([u8; panic!(1)]); //~ ERROR panicking in constants is unstable
   |
   = note: see issue #51999 <https://github.com/rust-lang/rust/issues/51999> for more information
   = help: add `#![feature(const_panic)]` to the crate attributes to enable
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: argument to `panic!()` in a const context must have type `&str`
   |
   |
LL | struct Bug([u8; panic!(1)]); //~ ERROR panicking in constants is unstable
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors
---

Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-9/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "9.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver engine executionengine fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:15:59
```

</details>

abonander commented:
I went ahead and made sure to cover `const fn` as well.

RalfJung commented:
> @RalfJung @oli-obk I adjusted the wording, fixed the check and added UI tests. I noticed a discrepancy between panics in array length expressions and normal consts so I put them in separate files. Should the tests cover `const fn` as well?

You can have more than one `//~ ERROR` in a single file, and that is usually preferred to group testcases.

abonander commented:
@RalfJung well the problem I had with arrays is that it seems their length expressions get const-evaluated in an earlier pass than `const` or `static` or `const fn` because the compiler would issue an error for `let _ = [0i32; panic!(1)]` and then exit before even visiting the other declarations. I assumed this was to be expected since arrays need to know their lengths for typechecking to finish, thus I separated the test cases into two files. Otherwise, I did combine test cases where I could.

RalfJung commented:
That's a good point, thanks. Maybe add a comment in the array-length-test-file explaining why this is a separate file.

abonander commented on src/test/ui/consts/issue-66693-panic-in-array-len.rs:
```diff
@@ -0,0 +1,17 @@
+// This is a separate test from `issue-66693.rs` because array lengths are evaluated
+// in a separate stage before `const`s and `statics` and so the error below is hit and
+// the compiler exits before generating errors for the others.
```
> That's a good point, thanks. Maybe add a comment in the array-length-test-file explaining why this is a separate file.

@RalfJung already done.

RalfJung replied:
Okay. :) I'll review this when I find some more time.

RalfJung commented on src/test/ui/consts/issue-66693-panic-in-array-len.rs:
```diff
@@ -0,0 +1,17 @@
+    let _ = [false; panic!()];
+    //~^ ERROR: evaluation of constant value failed
+
+    // const-evaluation halts before getting to this one
```
```suggestion
    // type checking halts before getting to this one
```

RalfJung commented on src/test/ui/consts/issue-66693.rs:
```diff
@@ -0,0 +1,25 @@
+}
+
+// ensure that conforming panics are still handled correctly
+// const-evaluation halts before triggering these but not in the array-len case
```
I do not understand the second part of this comment... it seems more confusing than helpful?

JohnCSimon commented:
ping from triage:
@abonander - can you address the comment from RalfJung ?

@Rustbot label: -S-waiting-on-review +S-waiting-on-author

JohnCSimon commented:
ping again from triage:
@abonander - can you address the comment from RalfJung?

SergioBenitez commented:
It seems like this is waiting on rather minor changes, is that right? @RalfJung I'm happy to push this forward if so.

RalfJung commented:
@SergioBenitez I left some comments that are all minor, yes (see the open discussions). I don't know if @oli-obk has further comments; he is more familiar with the const-checking code.

abonander commented:
@RalfJung sorry, been busy with work. I think that's your nits addressed now.

@JohnCSimon if you want to update the triage label.

abonander commented:
@rustbot label: +S-waiting-on-review -S-waiting-on-author

oli-obk commented on compiler/rustc_mir/src/transform/check_consts/validation.rs:
```diff
@@ -857,9 +857,18 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
+                    // const-eval of the `begin_panic` fn assumes the argument is `&str`
+                    if Some(callee) == tcx.lang_items().begin_panic_fn() {
+                        match args[0].ty(&self.ccx.body.local_decls, tcx).kind() {
+                            ty::Str => (),
```
This arm cannot happen, string slices are always behind a reference

oli-obk reviewed (commented):
Just a nit, then this lgtm

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 5a33f531cd87988e95a8e811d71b653a7ff9ffca has been approved by `oli-obk`

<!-- @bors r=oli-obk 5a33f531cd87988e95a8e811d71b653a7ff9ffca -->
<!-- homu: {"type":"Approved","sha":"5a33f531cd87988e95a8e811d71b653a7ff9ffca","approver":"oli-obk"} -->
